### DISCUSSION TITLE
[Announcement] Tailwind CSS v1.4 is out, with built-in PurgeCSS support!

### DISCUSSION THREAD

**Discussion by adamwathan** [tailwindlabs/tailwindcss#1656] [General]
Tailwind CSS v1.4.0 is here! 🥳  Another new minor version so soon? Merry Coronavirus 🎄😷🎄

## New Features

- [New color opacity utilities](#new-color-opacity-utilities)
- [Built-in PurgeCSS](#built-in-purgecss)
- [IE 11 target mode](#ie-11-target-mode) (experimental)

<a name="new-space-and-divide-layout-utilities"></a>

### New color opacity utilities (\#1627)

Tailwind v1.4 adds a new set of utilities for controlling just the alpha channel of colors:

- `bg-opacity-{value}`
- `text-opacity-{value}`
- `border-opacity-{value}`
- `divide-opacity-{value}`
- `placeholder-opacity-{value}`

These utilities compose with the existing color utilities like this:

```html
<div class="bg-red-500 bg-opacity-25">
  <!-- ... -->
</div>
```

All of these new utilities inherit their values from the `opacity` config by default but can be configured independently under the following theme keys:

- `backgroundOpacity`
- `textOpacity`
- `borderOpacity`
- `placeholderOpacity`
- `divideOpacity`

Learn more in the [pull request](https://github.com/tailwindcss/tailwindcss/pull/1627).

<a name="built-in-purgecss"></a>

### Built-in PurgeCSS (\#1639)

Tailwind v1.4 adds a new `purge` option to the config for removing unused CSS without the need for configuring another tool:

```js
// tailwind.config.js
module.exports = {
  purge: [
    './src/**/*.html',
    './src/**/*.vue',
    './src/**/*.jsx',
  ],
  theme: {},
  variants: {},
  plugins: [],
}
```

Learn more in the [pull request](https://github.com/tailwindcss/tailwindcss/pull/1639) and the [updated "Controlling File Size" documentation](https://tailwindcss.com/docs/controlling-file-size/).

<a name="ie-11-target-mode"></a>

### IE 11 target mode (\#1635)

**This feature is experimental and may change outside of normal semantic versioning policies.**

Recently we've been adding a lot of new features to Tailwind that aren't IE 11 compatible, and if you need to support IE 11 in your projects you have to be careful not to rely on these features by mistake.

To make this easier, we've added a new `target` option to the config that lets you opt-in to a new `ie11` mode that disables any features that are not compatible with IE 11:

```js
// tailwind.config.js
module.exports = {
  target: 'ie11',
  theme: {},
  variants: {},
  plugins: [],
}
```

Learn more in the [pull request](https://github.com/tailwindcss/tailwindcss/pull/1635).

**Comment by udemezue01** [comment]
Good job sir
we need a modal component.

**Reply by nunomaduro** [reply]
Here is one: https://tailwindui.com/components/application-ui/overlays/modals.

**Comment by vivgui** [comment]
I think the **Built-in PurgeCSS** feature is a complete game-changer, no need to fiddle with webpack/gulp config, or deal with a third party implementation's purging decisions (I'm looking at you [@nuxtjs/tailwindcss](https://github.com/nuxt-community/tailwindcss-module)) anymore.

**Reply by wpsix** [reply]
It's fine but wasn't so hard to setup.

**Comment by stuyam** [comment]
I can't tell if it is just me or if the `bg-opacity-0` class is not working on the docs pages. 
![image](https://user-images.githubusercontent.com/1553779/80665833-86bd3480-8a68-11ea-8872-d524cb006f27.png)
Far right one should be completely transparent if I am not mistaken? Not sure if this is an issue in the docs or the main package.

**Reply by ajitbohra** [reply]
@adamwathan seems like `.bg-opacity-100` & `.bg-opacity-0` both are purged from final styles on docs site.

Available on CDN copy and works fine.

**Reply by MZanggl** [reply]
Interesting issue as it shows the dangers of purging only on production (which is the default setting). The same thing happened to me the other day with nuxt, because their plugin purged classes in `nuxt.config.js`.

I wonder if there are some possible improvements to make purging during development more performant to avoid these surprises.

**Reply by adamwathan** [reply]
Fixed, problem is I break my own rules on the Tailwind site and do some stupid things with dynamic classes and loops, because otherwise trying to maintain the docs would be a total nightmare. Thanks for the heads up!

**Comment by benoliver999** [comment]
Thank you for the built in purge, it is a crucial step and very handy.

Is there a way to exclude a directory from ` purge: [` ?

I am using jekyll with postcss, and it would be good to exclude `_site`, in order to avoid running purge on an old build.

**Reply by kevinruscoe** [reply]
As per https://tailwindcss.com/docs/controlling-file-size/#purgecss-options, you can pass an `options` object straight to `PurgeCSS`. So, in that case, check out the [purgeCSS docs](https://purgecss.com/configuration.html#options) on how to change paths, specifically the `content` option.

**Comment by janko** [comment]
Are you open to extending the default purge regex to support other templating languages? For my HAML templates I found that I needed to add `=` to support this common syntax:

```haml
.text-gray-500= content
```

This change worked for me (I overrode both just in case):

```diff
- const broadMatches = content.match(/[^<>"'`\s]*[^<>"'`\s:]/g) || []
- const innerMatches = content.match(/[^<>"'`\s.()]*[^<>"'`\s.():]/g) || []
+ const broadMatches = content.match(/[^<>"'`\s]*[^<>"'`\s:=]/g) || []
+ const innerMatches = content.match(/[^<>"'`\s.()]*[^<>"'`\s.():=]/g) || []
```

**Reply by utkarshkukreti** [reply]
@adamwathan maybe a whitelist of allowed characters would be better than a blacklist to support these templating languages? `[a-zA-Z0-9:\/_-]` and probably some I've missed.

**Reply by adamwathan** [reply]
I was using a whitelist originally but it was annoying to always be adding new characters, for example Tailwind UI uses the `.` character in classes, and the Tailwind website itself uses `(` and `)` in classes. I think adding `=` to the innerMatches pass is a good solution, basically we just want to track possible delimiters rather than possible valid characters 👍 

Will update and tag a patch soon.

**Reply by utkarshkukreti** [reply]
👍  maybe add `[]` and `{}` as well. Haml lets you use `{}` and Slim `[]` for attributes, e.g.
`.a[b=1]` compiles to `<div class='a' b=1>` in Slim.

**Reply by adamwathan** [reply]
Ah good to know, adding that as well!

**Reply by janko** [reply]
Thank you for such a prompt response ❤️ Yes, both of your suggestions sound good to me 👍

**Reply by adamwathan** [reply]
v1.4.1 is out with better support for Haml and Slim! 👍

**Reply by janko** [reply]
Wow, thank you for such a quick patch and release. I've just upgraded and it works marvelously 👌

**Comment by kevinruscoe** [comment]
FYI, the `target: 'ie11',` is added for quickly targetting IE11. The PR for this specifically mentions using browserslist (thank god) in the future, albeit a slimed, sensible version.

**Comment by glauberfc** [comment]
The integration with PurgeCSS was a great decision!

**Comment by stefanstifter** [comment]
also very excited about the integrated purgeCSS. can't wait to try that out. thanks!

**Comment by dixonge** [comment]
If using PostCSS, upgrading TailwindCSS means I need to move all of the PurgeCSS plugin functions over to the new 'purge' option inside of Tailwind plugin? I am unclear on what to move where. :/

**Reply by devcircus** [reply]
You can choose to keep using purgecss as you currently do. No reason to move anything if you don't want to.

**Comment by adamwathan** [comment]
@dixonge Can you share your current PostCSS config? Happy to tell you what to put where.

**Reply by dixonge** [reply]
```
const purgecss = require("@fullhuman/postcss-purgecss")({
  // Specify the paths to all of the template files in your project
  content: ["./dist/**/*.html"],
  whitelist: ["dark-mode"],

  // Include any special characters you're using in this regular expression
  defaultExtractor: (content) => content.match(/[\w-/:%.]+(?<!:)/g) || [],
});

module.exports = {
  plugins: [
    require("tailwindcss"),
    require("autoprefixer"),
    ...(process.env.NODE_ENV === "production"
      ? [purgecss, require("cssnano")]
      : []),
  ],
};
```
It is using cssnano with purgecss...

**Reply by adamwathan** [reply]
This should do it for you:

```js
// postcss.config.js
module.exports = {
  plugins: [
    require("tailwindcss"),
    require("autoprefixer"),
    ...(process.env.NODE_ENV === "production"
      ? [require("cssnano")]
      : []),
  ],
};

// tailwind.config.js
module.exports = {
  purge: {
    content: ["./dist/**/*.html"],
    whitelist: ["dark-mode"],
  }
}
```

**Reply by dixonge** [reply]
Yes sir! Thank you so much...

**Comment by samanthaming** [comment]
This is awesome! The `Built-in PurgeCSS` is probably what I'm most excited about -- as I was looking into installing it separately, but now I don't have to 👏

**Question**: My site is built with VuePress, is there any additional setup/configuration I need to do to get it to work? Or would everything just magically work 😄
